File: Connect4/algorithms/minimax/successor_functions/Spiral.cpp
/*
 * Spiral.cpp
 *
 *  Created on: Feb 20, 2018
 *      Author: JLepere2
 */

#include "Spiral.h"

vector<vector<int> > Spiral::execute(vector<vector<int> >& board, int previousRow, int previousCol)
{
	vector<vector<int> > successors;

	int r;
	int c;
	int h;
	int x = max(max(previousRow, BOARD_SIZE - previousRow - 1), max(previousCol, BOARD_SIZE - previousCol - 1));
	for (int d = 1; d <= x; d ++) {
		r = previousRow - d;
		c = previousCol + d;
		h = d * 2;
		// DOWN
		if ((c >= 0) && (c < BOARD_SIZE)) {
			for (int i = 0; i < h; i ++) {
				r += 1;
				if ((r >= 0) && (r < BOARD_SIZE)) {
					if (board[r][c] == 0) {
						successors.push_back({r,c});
					}
				}
			}
		} else {
			r = previousRow + d;
		}
		// LEFT
		if ((r >= 0) && (r < BOARD_SIZE)) {
			for (int i = 0; i < h; i ++) {
				c -= 1;
				if ((c >= 0) && (c < BOARD_SIZE)) {
					if (board[r][c] == 0) {
						successors.push_back({r,c});
					}
				}
			}
		} else {
			c = previousCol - d;
		}
		// UP
		if ((c >= 0) && (c < BOARD_SIZE)) {
			for (int i = 0; i < h; i ++) {
				r -= 1;
				if ((r >= 0) && (r < BOARD_SIZE)) {
					if (board[r][c] == 0) {
						successors.push_back({r,c});
					}
				}
			}
		} else {
			r = previousRow - d;
		}
		// RIGHT
		if ((r >= 0) && (r < BOARD_SIZE)) {
			for (int i = 0; i < h; i ++) {
				c += 1;
				if ((c >= 0) && (c < BOARD_SIZE)) {
					if (board[r][c] == 0) {
						successors.push_back({r,c});
					}
				}
			}
		}
	}

	return successors;
}

File: MiniMaxJava/src/successors/Spiral.java
package successors;

import java.util.ArrayList;

import cs420_p3.State;

public class Spiral implements Successors {

	/**
	 * We will go right, up, left, down, and then right.
	 */
	@Override
	public int[][] getSuccessors(int[][] board, int row, int col) {
		ArrayList<int[]> temp = new ArrayList<int[]>();
		
		int increasedBy = 1;
		int boardSize = State.BOARD_SIZE;
		int maxSearch = (boardSize - 1) * (boardSize - 1);
		boolean successorsRemain = true;
		
		while (successorsRemain) {
			int rowBoundary, colBoundary, curRow, curCol;
			
			if ((col + increasedBy) <= maxSearch || (row - increasedBy) >= 0 ||
					(col - increasedBy) >= 0 || (row + increasedBy) <= maxSearch) {
				// Top right
				colBoundary = col + increasedBy;
				rowBoundary = row - increasedBy;
				curRow = row;
				curCol = colBoundary;
				for (int i = 0; i < increasedBy+1; i++) {					
					if (curRow >= 0 && curRow < boardSize && curCol >= 0 && curCol < boardSize) {
						if (board[curRow][curCol] == 0) {
							temp.add(new int[]{curRow, curCol});
						}
					}
					curRow--;
					curCol--;
				}
				
				// Top left
				colBoundary = col - increasedBy;
				curRow = row;
				curCol = colBoundary;
				for (int i = 0; i < increasedBy; i++) {					
					if (curRow >= 0 && curRow < boardSize && curCol >= 0 && curCol < boardSize) {
						if (board[curRow][curCol] == 0) {
							temp.add(new int[]{curRow, curCol});
						}
					}
					curRow--;
					curCol++;
				}
				
				// Bottom left
				rowBoundary = row + increasedBy;
				curRow = rowBoundary;
				curCol = col;
				for (int i = 0; i < increasedBy; i++) {
					if (curRow >= 0 && curRow < boardSize && curCol >= 0 && curCol < boardSize) {
						if (board[curRow][curCol] == 0) {
							temp.add(new int[]{curRow, curCol});
						}
					}
					curRow--;
					curCol--;
				}
				
				// Bottom right
				colBoundary = col + increasedBy;
				curRow = row + increasedBy - 1;
				curCol = col + 1;
				for (int i = 0; i < increasedBy-1; i++) {					
					if (curRow >= 0 && curRow < boardSize && curCol >= 0 && curCol < boardSize) {
						if (board[curRow][curCol] == 0) {
							temp.add(new int[]{curRow, curCol});
						}
					}
					curRow--;
					curCol++;
				}
				
				increasedBy++;
			}
			else {
				successorsRemain = false;
			}
		}
		
		int[][] returnValue = new int[temp.size()][2];
		for (int i = 0; i < returnValue.length; i++) {
			for (int j = 0; j < returnValue[i].length; j++) {
				returnValue[i][j] = temp.get(i)[j];
			}
		}
		
		return returnValue;
	}
}